PLANNER-536 Workbench solver editor: bendable score combobox selection should allow filling in hard and soft level size

mcimbora opened the pull request:
Depends on https://github.com/droolsjbpm/kie-wb-common/pull/466

Adds full support for using bendable score types in `@PlanningSolution` data objects.

Commit:
PLANNER-536 Workbench solver editor: bendable score combobox selection should allow filling in hard and soft level size

wmedvede commented on optaplanner-wb-screens/optaplanner-wb-domain-editor/optaplanner-wb-domain-editor-backend/src/main/java/org/optaplanner/workbench/screens/domaineditor/backend/server/PlannerMethodFilter.java:
```diff
@@ -0,0 +1,32 @@
+
+    @Override
+    public boolean accept( Method<?, ?> method ) {
+        return !method.isConstructor() && method.getAnnotation( Generated.class ) != null;
```
Hello @mcimbora , I think it can be worth to provide a more fine grained filtering here. To be sure that the method  is not a constructor, has the @Generated annotation, and also "is one of the generated" by the planner stuff. Probably inspecting some additional planner generated annotation for the method (I I don't remember wrong for this planner generated method you added has also a planner annotation with two parameters?) or eventually exploring the return type, or some other trick. This will prevent us from loading other methods with the @Generated annotation that we don't know how to manage. Specially because when the java source is being updated, all the accepted methods are being deleted.

mcimbora commented:
Jenkins retest this

mcimbora commented:
@wmedvede @csadilek I updated the PR to reflect the changes from https://github.com/droolsjbpm/kie-wb-common/pull/466.

wmedvede commented:
Hello @mcimbora , regarding the org.optaplanner.workbench.screens.domaineditor.model.ComparatorDefinition annotation I don't know exactly from the planner perspective what's the best suited package. But likelly it should go in a package  like (just a suggestion) like org.optaplanner.core.api.model. This is basically for avoiding build errors for clients that wants to build the generated class in their own IDE. Since it's more likelly that they imports the XXX.jar that contains all the planer api, but not the workbench.screen.domianeditor.model

mcimbora commented:
Hi @wmedvede I updated the PR. I did a lot of polishing around the comparator definition. I introduced the new definition scheme combining `@ComparatorDefinition`, `@ComparatorObjectPropertyPath` and `@ComparatorObjectProperty`. I moved processing of the comparator to the client side. Method/nested class filters have been added to make sure we load only WB-generated stuff.  As to the location of the `ComparatorDefinition` annotation, I'll discuss it with @ge0ffrey & let you know about the outcome. Thanks!

wmedvede commented on optaplanner-wb-screens/optaplanner-wb-domain-editor/optaplanner-wb-domain-editor-client/src/main/java/org/optaplanner/workbench/screens/domaineditor/client/handlers/planner/PlannerDomainHandler.java:
```diff
@@ -78,193 +79,180 @@ public void postCommandProcessing( DataModelCommand command ) {
+                                        Annotation planningEntityAnnotation = dataObject.getAnnotation( PlannerDomainAnnotations.PLANNING_ENTITY_ANNOTATION );
+                                        planningEntityAnnotation.setValue( "difficultyComparatorClass", dataObject.getName() + "." + dataObject.getName() + "Comparator.class" );
+
+                                        dataModelerService.call().saveSource( event.getSource(), event.getPath(), dataObject, null, "Update comparator object after data object class name change" );
```
I'm afraid that we can't save the java file from this point independently from users decision. If we do this we can leave the file inconsistent. e.g. I can imagine a scenario  when the user opens the file, plays a bit (changing eventually the class name, etc.) and then closes the file by cancelling the modifications in the UI. If the file was automatically saved it will be inconsistent with the user desire.

mcimbora replied:
@wmedvede The decision was based on the fact that I couldn't find any occurrences of CLASS_NAME_CHANGE in the project. I reused the change type in the event fired when the file (using 'Rename') is renamed https://github.com/droolsjbpm/kie-wb-common/pull/466/files#diff-5166e0a37add3f6506d1c15648308d73R539. Perhaps I can introduce separate event here, e.g. `DataObjectRenamedEvent` if that makes more sense?. The reason the event is needed is that the handler needs to react appropriately to name changes (e.g. rename the nested comparator class, change parameter types).

mcimbora commented:
@wmedvede I updated the PR. The comparator annotations will remain in the workbench for now, in the upcoming weeks it will be added to the optaplanner-core as well, however the discussion is still ongoing how the final state will look like. https://issues.jboss.org/browse/PLANNER-667

wmedvede commented on optaplanner-wb-screens/optaplanner-wb-domain-editor/optaplanner-wb-domain-editor-client/src/main/java/org/optaplanner/workbench/screens/domaineditor/client/handlers/planner/PlannerDomainHandler.java:
```diff
@@ -32,29 +35,31 @@
 public class PlannerDomainHandler implements DomainHandler {
 
-    private static final Logger logger = LoggerFactory.getLogger( PlannerDomainHandler.class );
+    private Caller<DataModelerService> dataModelerService;
```
Hello @mcimbora , I believe this dataModelerService is no longer used.

wmedvede commented:
Hello @mcimbora  @csadilek , +1 from my side.
